lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -287,6 +288,26 @@ protected void readField(Object record, Field field, Object oldDatum, ResolvingD
             return;
           }
         }
+
+        if (field.schema().isUnion()) {
```

Comment:
You've pulled handling the union up from `readWithoutConversion`.
This means that handling fields like this works, but if the dates are in an array of unions, it will fail (see the method `readArray` starting at line 125).

It's probably better to put this logic in an overridden version of `GenericDatumReader#read(Object, Schema, ResolvingDecoder)`, which is called from `readWithoutConversion`. Granted, this applies conversion inside a method named "without conversion", but _the current implementation already does (buggy) conversions for unions_, so we won't be off worse.

Reply:
Hi opwvhk, thanks for reviewing. 

In order to get correct conversion, I need both class type and logical type (since two different classes are mapped to the same logical type). At line 298, I can get the class type using `accessor.getField().getType()` and pass it to `getData().getConversionByClass`. I guess for the same reason, the logic of line 277-290 also ended up where it is now (see line 283 where it passes `accessor.getField().getType()` to `getData().getConversionByClass`).

If the logic is moved to `GenericDatumReader#read(Object, Schema, ResolvingDecoder)` then we _cannot_ get class type without passing an extra argument into the `GenericDatumReader#read` method. This may requires changing a lot of the code. 

I think `ReflectData.AllowNull` doesn't generate the array of unions since it only makes class fields nullable (see: https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java#L100), and therefore the logic is put into `ReflectDatumReader#readField` method.

Reply:
Actually, the method `getAnnotatedUnion` has 4 usages. This method allows creating unions for common supertypes, and is the reason unions also occur for other use cases than nullable fields.

As a result, I think we do need to override that `GenericDatumReader#read` method in `ReflectDatumReader`. Maybe by directly calling an overloaded version with an extra type parameter (`null` in the overridden `read` method as it's unknown).

Reply:
The only problem will be how this extra parameter will spread. We need to come up with a way to prevent that.

Maybe we should add a schema property to generated unions to store the target type?

Reply:
This just occurred to me: maybe we should bypass this union issue by adding the target type property to the schema with the logical type. After all, the underlying cause is that some conversions (most commonly date/time conversions) have multiple target types.

Anyway, have a look at the callsites of `makeNullable` and `getAnnotatedUnion` and choose the best/easiest option.

Reply:
I like the idea of adding target type property to the schema and overriding the `GenericDatumReader#read` method in `ReflectDatumReader`. I tried it by adding a property `SpecificData.CLASS_PROP` (java-class) indicating java class for logical types (it seems CLASS_PROP is the conventional property name used to indicate java class in the code base). It broke a few unit tests but they're easy to fix. 

It almost worked but when I tried fixing the last broken unit test `TestReflectLogicalTypes.testReflectedSchemaLocalDateTime`, I got the following exception: 
![image](https://user-images.githubusercontent.com/34998797/148373820-9d31671d-d024-496d-a94e-a1b4477d369e.png)

Although I was able to add java-class property to schema with a logical type, it seems some code assumes logical types property should not be used with java-class. PS: The code changes are here, though I'm no longer sure we should go this way: https://github.com/liu-du/avro/commit/6320aae25b6f3c8f10d48b6b4e0fd31d03ebd1bd .